bin/vcfparser.py

```diff
@@ -0,0 +1,180 @@
+                            to the INFO column
+    """
+
+    columns = line.strip().split()
+    transcripts = columns[7].split("CSQ=")[1].split(",")
+    hgnc_ids, alleles, consequences, severity = parse_vep_csq_transcripts(
+        transcripts, allele_ind, csq_ind, hgnc_ind, var_csq
+    )
+    unique_ids = list(set(hgnc_ids))
+    mscsq_anno = []
+    for gene_id in unique_ids:
+        if gene_id != "":
+            indices = find_indices(hgnc_ids, gene_id)
+            alleles_sub = [alleles[i] for i in indices]
+            consequences_sub = [consequences[i] for i in indices]
+            severity_sub = [severity[i] for i in indices]
+            most_severe_csq = consequences_sub[severity_sub.index(min(severity_sub))]
+            most_severe_allele = alleles_sub[severity_sub.index(min(severity_sub))]
+            mscsq_anno.append(gene_id + ":" + most_severe_allele + "|" + most_severe_csq)
+    columns[7] += ";most_severe_consequence=" + ",".join(mscsq_anno)
+    return columns
+
+
+def find_indices(list_to_check, item_to_find):
+    """
+    Get indices of an element in a list
+
+    Args:
+        list_to_check (list)
+        item_to_find  (value)
+
+    Returns:
+        indices       (list)
+    """
+    indices = []
+    for idx, value in enumerate(list_to_check):
+        if value == item_to_find:
+            indices.append(idx)
+    return indices
+
+
+def parse_vep_csq_schema(line):
+    """
+    Get indices of allele, consequence, and hgnc id in the annotation
+
+    Args:
+        line: INFO line in the vcf header with CSQ information
+
+    Returns:
+        allele_ind  (int) : Index of the "allele" in the vep annotation record
+        csq_ind     (int) : Index of the "Consequence" in the vep annotation record
+        hgnc_ind    (int) : Index of the "HGNC_ID" in the vep annotation record
+    """
+    fields = line.strip().split("Format: ")[1].replace('">', "").split("|")
+    allele_ind = fields.index("Allele")
+    csq_ind = fields.index("Consequence")
+    hgnc_ind = fields.index("HGNC_ID")
+
+    return allele_ind, csq_ind, hgnc_ind
+
+
+def write_csq_annotated_vcf(file_in, file_out, var_csq):
+    """Add most severe consequence field to record, and write the record to a vcf file"""
+    counter = 0
+    for line in file_in:
+        if line.startswith("#"):
+            if line.startswith("##INFO=<ID=CSQ"):
+                counter = 1
+                allele_ind, csq_ind, hgnc_ind = parse_vep_csq_schema(line)
+            elif not line.startswith("##INFO") and counter == 1:
+                file_out.write(
+                    '##INFO=<ID=most_severe_consequence,Number=.,Type=String,Description="Most severe genomic consequence.">\n'
+                )
+                counter = 0
+            file_out.write(line)
+        else:
+            mscsq = construct_most_severe_consequence_info(line, allele_ind, csq_ind, hgnc_ind, var_csq)
+            file_out.write("\t".join(mscsq) + "\n")
```

Comment:
````python 
def write_csq_annotated_vcf(file_in, file_out, var_csq):
    """Add most severe consequence field to record, and write the record to a vcf file"""
    for line in file_in:
        if line.startswith("#"):
            file_out.write(line)
            if line.startswith("##INFO=<ID=CSQ"):
                allele_ind, csq_ind, hgnc_ind = parse_vep_csq_schema(line)
                file_out.write(
                    '##INFO=<ID=most_severe_consequence,Number=.,Type=String,Description="Most severe genomic consequence.">\n'
                )
        else:
            mscsq = construct_most_severe_consequence_info(line, allele_ind, csq_ind, hgnc_ind, var_csq)
            file_out.write("\t".join(mscsq) + "\n")
````
Suggestion that I can think makes it possible to remove an `elif` statement

Reply:
Good that you brought this up, MIP's vcfparser adds this line to the end of the INFO block so I tried to replicate it. But I agree that it doesn't matter as long as the key is in there.